python/lib/modeldata/dmod/modeldata/data/object_store_dataset.py

```diff
@@ -0,0 +1,369 @@
+        in their ::method:`__init__`.  This function should do this deserialization work for any subtype, and return a
+        deserialized dictionary.  The keys should be the names of the relevant ::method:`__init__` parameters.
+
+        In the event a type's ::method:`__init__` method takes no additional params beyond the base type, its
+        implementation of this function should return an empty dictionary.
+
+        Any types with an init that does not have one or more of the params of the base type's init should fully
+        override ::method:`factory_init_from_deserialized_json`.
+
+        Parameters
+        ----------
+        json_obj : dict
+            The serialized form of the object that is a subtype of ::class:`Dataset`.
+
+        Returns
+        -------
+        Dict[str, Any]
+            A dictionary of ``kwargs`` for those init params and values beyond what the base type uses.
+        """
+        # TODO: update docstring a bit for this once finalized nothing is needed
+        return dict()
+
+    def __init__(self, name: str, category: DataCategory, data_domain: DataDomain, access_location: str,
+                 uuid: Optional[UUID] = None, manager: Optional[DatasetManager] = None,
+                 manager_uuid: Optional[UUID] = None, is_read_only: bool = True, expires: Optional[datetime] = None,
+                 derived_from: Optional[str] = None, derivations: Optional[List[str]] = None,
+                 created_on: Optional[datetime] = None, last_updated: Optional[datetime] = None):
+        super(ObjectStoreDataset, self).__init__(name, category, data_domain, access_location, uuid, manager,
+                                                 manager_uuid, is_read_only, expires, derived_from, derivations,
+                                                 created_on, last_updated)
+        # TODO: remove this if there ends up being no use for it
+
+    def add_file(self, file: Path, add_relative_to: Optional[Path] = None) -> bool:
+        """
+
+        Parameters
+        ----------
+        file
+        add_relative_to
+
+        Returns
+        -------
+        bool
+            Whether the file was added successfully
+        """
+        bucket_root = file.parent if add_relative_to is None else add_relative_to
+        return self.manager.add_data(dataset_name=self.name, file=file, bucket_root=bucket_root)
+
+    def add_files(self, directory: Path, add_relative_to: Optional[Path] = None) -> bool:
+        """
+        Add all files in the given directory to this existing object store.
+
+        Parameters
+        ----------
+        files : List[Path]
+            The files to be added
+        """
+        bucket_root = directory if add_relative_to is None else add_relative_to
+        return self.manager.add_data(dataset_name=self.name, directory=directory, bucket_root=bucket_root)
+
+    @property
+    def files(self) -> List[str]:
```

Comment:
Hmm ...  

Objects in the bucket correspond to files, but they are not really files at that point.  So it may not always be appropriate to act on them like files.  It also it doesn't seem like it would be necessary or useful (without some other things happening first).   

It does still seem useful to "think" of them as files, though; i.e., being able to get information on the structure of the "files" in the dataset may be useful, without/before actually having handles to the files.  

Plus, a user could trivially obtain `Path` objects by using the strings.  That would probably be better because the user could then control the process and make sure the `Path` object were  valid with respect to a relative root.

Reply:
The point of a `Path` is to be a path, doesn't _have_ to be an existing file/dir, just a path with path conventions.  Not a deal breaker on this PR, though.